src/main/scala/org/constellation/ConstellationNode.scala

```diff
@@ -198,24 +204,32 @@ class ConstellationNode(
 
   // Setup http server for peer API
   // TODO: Add shutdown mechanism
-  Http()
+  val peerApiBinding: Future[Http.ServerBinding] = Http()
     .bind(nodeConfig.httpInterface, nodeConfig.peerHttpPort)
-    .runWith(Sink.foreach { conn =>
+    .to(Sink.foreach { conn =>
       val address = conn.remoteAddress
       conn.handleWith(peerAPI.routes(address))
     })
+    .run()
 
   def shutdown(): Unit = {
-    val gracefulShutdown = IO.delay(bindingFuture.foreach(_.unbind())(ConstellationExecutionContext.callbacks)) >>
-      IO.delay(logger.info("Node shutdown completed"))
 
-    dao.cluster
-      .leave(gracefulShutdown)
-      .unsafeRunSync()
+    val unbindTimeout = ConfigUtil.getDurationFromConfig("akka.http.unbind-api-timeout")
 
-    // TODO: we should add this back but it currently causes issues in the integration test
-    //.onComplete(_ => system.terminate())
+    implicit val ec: ExecutionContextExecutor = ConstellationExecutionContext.unbounded
 
+    val gracefulShutdown: IO[Terminated] = IO.fromFuture(
+      IO(
```

Comment:
If you're using `IO`, you should wrap each part into single `IO.delay`, so one for terminating peerAPI, second one for terminating API, and third one for system.terminate.
https://typelevel.org/cats-effect/datatypes/io.html#keep-granularity